src/kontrol/prove.py

```diff
@@ -978,101 +987,104 @@ def _create_cse_accounts(
+                address_range_ub = ltInt(field_variable, intToken(1461501637330902918203684832716283019655932542976))
+                new_account_constraints.append(mlEqualsTrue(address_range_lb))
+                new_account_constraints.append(mlEqualsTrue(address_range_ub))
+        # Processing of contracts
         if field.data_type.startswith('contract '):
             contract_type = field.data_type.split(' ')[1]
             for full_contract_name, contract_obj in foundry.contracts.items():
                 # TODO: this is not enough, it is possible that the same contract comes with
                 # src% and test%, in which case we don't know automatically which one to choose
                 if full_contract_name.split('%')[-1] == contract_type:
                     contract_account_code = bytesToken(bytes.fromhex(contract_obj.deployed_bytecode))
-                    contract_account_name = 'CONTRACT-' + field.label.upper()
-
-                    # TODO: handle the case where the field has a non-zero offset
-                    # maxUInt160 &Int #lookup ( CONTRACT_STORAGE:Map , 0 )
-                    contract_storage_lookup = KApply(
-                        '_&Int_',
-                        [
-                            intToken(1461501637330902918203684832716283019655932542975),
-                            KEVM.lookup(storage_map, intToken(field.slot)),
-                        ],
-                    )
-                    new_account_constraints.append(
-                        mlEqualsTrue(
-                            KApply('_==Int_', [contract_storage_lookup, KVariable(contract_account_name + '_ID')])
-                        )
-                    )
+                    contract_account_variable = KVariable(field_name + '_ID', sort=KSort('Int'))
 
                     # New contract account is not the cheatcode contract
                     new_account_constraints.append(
                         mlEqualsFalse(
                             KApply(
                                 '_==Int_',
                                 [
-                                    KVariable(contract_account_name + '_ID', sort=KSort('Int')),
+                                    contract_account_variable,
                                     Foundry.address_CHEATCODE(),
                                 ],
                             )
                         )
                     )
+                    # The contract address must occupy the entire slot
+                    if field.slot in singly_occupied_slots:
+                        # The offset must equal zero
+                        assert field.offset == 0
+                        # Update storage map accordingly: ( field.slot |-> contract_account_variable ) STORAGE_MAP
+                        storage_map = KApply(
+                            '_Map_', [map_item(intToken(field.slot), contract_account_variable), storage_map]
+                        )
```

Comment:
This part seems duplicated, could you extract it in an auxiliary function?

Reply:
Done.